R/summary.R

```diff
@@ -0,0 +1,220 @@
+#'   lower = gs_b,
+#'   lpar = rep(-Inf, 3))
+#'
+#' # Define cuttings of 2 IAs and 1 FA
+#' ia1_cut <- create_cut(target_event_overall = ceiling(design$analysis$event[1]))
+#' ia2_cut <- create_cut(target_event_overall = ceiling(design$analysis$event[2]))
+#' fa_cut <- create_cut(target_event_overall = ceiling(design$analysis$event[3]))
+#'
+#' # Run simulations
+#' simulation <- sim_gs_n(
+#'   n_sim = 3,
+#'   sample_size = ceiling(design$analysis$n[3]),
+#'   enroll_rate = design$enroll_rate,
+#'   fail_rate = design$fail_rate,
+#'   test = wlr,
+#'   cut = list(ia1 = ia1_cut, ia2 = ia2_cut, fa = fa_cut),
+#'   weight = fh(rho = 0, gamma = 0.5))
+#'
+#' # Summarize simulations
+#' simulation |> summary(bound = gsDesign::gsDesign(k = 3, test.type = 1, sfu = gsDesign::sfLDOF)$upper$bound)
+#'
+#' # Summarize simulation and compare with the planned design
+#' simulation |> summary(design = design)
+summary.simtrial_gs_wlr <- function(object,
+                                    design = NULL,
+                                    bound = NULL,
+                                    ...) {
+  # get the total number of analysis and simulations
+  n_analysis <- nrow(object[object$sim_id == 1, ])
+  n_sim <- nrow(object) / n_analysis
+
+  # if the design input is NULL
+  # then simply output the simulated n, event, power
+  if (is.null(design)) {
+    ans1 <- object |>
+      dplyr::group_by(analysis) |>
+      dplyr::summarize(sim_n = mean(n), sim_event = mean(event), sim_time = mean(cut_date))
+
+    ans2 <- object |>
+      dplyr::left_join(data.frame(analysis = 1:n_analysis, upper_bound = bound)) |>
+      dplyr::mutate(cross_upper = -z >= upper_bound) |>
+      dplyr::filter(cross_upper == TRUE) |>
+      dplyr::group_by(sim_id) |>
+      dplyr::filter(dplyr::row_number() == 1) |>
+      dplyr::ungroup() |>
+      dplyr::group_by(analysis) |>
+      dplyr::summarize(n_cross_upper = dplyr::n()) |>
+      dplyr::mutate(sim_upper_prob = cumsum(n_cross_upper) / n_sim) |>
+      dplyr::select(analysis, sim_upper_prob)
+
+    suppressMessages(
+      ans <- ans1 |> dplyr::left_join(ans2)
+    )
+
+    attr(ans, "compare_with_design") <- "no"
+  } else {
+    # get the design type, 1-sided or 2-sided
+    design_type <- ifelse(length(unique(design$bound$bound)) == 1, "one-sided", "two-sided")
+
+    # add the futility and efficacy bounds to the simulation results
+    if (design_type == "one-sided") {
+      suppressMessages(
+        sim_tbl <- object |>
+          dplyr::left_join(
+            design$bound |>
+              dplyr::select(analysis, z, bound) |>
+              dplyr::rename(upper_bound = z)
+          ) |>
+          dplyr::mutate(cross_upper = -z >= upper_bound)
+      )
+    } else {
+      suppressMessages(
+        sim_tbl <- object |>
+          dplyr::left_join(
+            design$bound |>
+              dplyr::select(analysis, z, bound) |>
+              tidyr::pivot_wider(values_from = z, names_from = bound) |>
```

Comment:
{dplyr} and {tidyr} are only in Suggests. Is the plan for me to convert this function to use {data.table}?

Reply:
Yes, please.